File: src/Tests/Behavioral/SelectStatement/SelectStatement.cs
namespace go;

using fmt = fmt_package;

partial class main_package {

[GoType("[]nint")] partial struct IntSlice;

internal static void g1(channel<nint> ch) {
    ch.ᐸꟷ(12);
}

internal static void g2(channel<nint> ch) {
    ch.ᐸꟷ(32);
}

internal static void sum(slice<nint> s, channel<nint> c) {
    nint sum = 0;
    foreach (var (_, v) in s) {
        sum += v;
    }
    c.ᐸꟷ(sum);
}

internal static void fibonacci(channel<nint> f, channel<nint> quit) {
    nint x = 0;
    nint y = 1;
    while (ᐧ) {
        var selᴛ1 = quit;
        switch (select(f.ᐸꟷ(x, ꓸꓸꓸ), ᐸꟷ(selᴛ1, ꓸꓸꓸ))) {
        case 0: {
            (x, y) = (y, x + y);
            break;
        }
        case 1 when selᴛ1.ꟷᐳ(out _): {
            fmt.Println("quit");
            return;
        }}
    }
}

internal static void sendOnly(channel/*<-*/<@string> s) {
    s.ᐸꟷ("output"u8);
}

public static Action<Func<nint, bool>> All(this IntSlice s) {
    var sʗ1 = s;
    return (Func<nint, bool> yield) => {
        foreach (var (_, v) in sʗ1) {
            if (!yield(v)) {
                return;
            }
        }
    };
}

internal static void generate(channel/*<-*/<nint> ch) {
    for (nint i = 2; ᐧ ; i++) {
        ch.ᐸꟷ(i);
    }
}

internal static void filter(/*<-*/channel<nint> src, channel/*<-*/<nint> dst, nint prime) {
    foreach (var i in src) {
        if (i % prime != 0) {
            dst.ᐸꟷ(i);
        }
    }
}

internal static void sieve() {
    var ch = new channel<nint>(0);
    goǃ(generate, ch);
    while (ᐧ) {
        nint prime = ᐸꟷ(ch);
        fmt.Print(prime, "\n");
        var ch1 = new channel<nint>(0);
        goǃ(filter, ch, ch1, prime);
        ch = ch1;
        if (prime > 40) {
            break;
        }
    }
}

internal static nint f() {
    return 0;
}

internal static void Main() {
    var ch = new channel<nint>(2);
    ch.ᐸꟷ(1);
    ch.ᐸꟷ(2);
    fmt.Println(ᐸꟷ(ch));
    fmt.Println(ᐸꟷ(ch));
    var a = new slice<nint>(2);
    var ch1 = new channel<nint>(0);
    var ch2 = new channel<nint>(0);
    var ch3 = new channel<nint>(0);
    var ch4 = new channel<nint>(0);
    goǃ(g1, ch1);
    goǃ(g2, ch2);
    goǃ(g1, ch3);
    goǃ(g2, ch4);
    for (nint i = 0; i < 4; i++) {
        var selᴛ2 = ch1;
        var selᴛ3 = ch2;
        var selᴛ4 = ch3;
        var selᴛ5 = ch4;
        switch (select(ᐸꟷ(selᴛ2, ꓸꓸꓸ), ᐸꟷ(selᴛ3, ꓸꓸꓸ), ᐸꟷ(selᴛ4, ꓸꓸꓸ), ᐸꟷ(selᴛ5, ꓸꓸꓸ))) {
        case 0 when selᴛ2.ꟷᐳ(out var v1): {
            fmt.Println("Got: ", v1);
            break;
        }
        case 1 when selᴛ3.ꟷᐳ(out var v1): {
            fmt.Println("Got: ", v1);
            break;
        }
        case 2 when selᴛ4.ꟷᐳ(out var v1, out var okΔ1): {
            fmt.Println("OK: ", okΔ1, " -- got: ", v1);
            break;
        }
        case 3 when selᴛ5.ꟷᐳ(out a[f()]): {
            fmt.Println("Got: ", a[f()]);
            break;
        }}
    }
    ch1 = default!;
    close(ch2);
    var selᴛ6 = ch1;
    var selᴛ7 = ch2;
    var selᴛ8 = ch3;
    var selᴛ9 = ch4;
    switch (select(ch1.ᐸꟷ(1, ꓸꓸꓸ), ᐸꟷ(selᴛ6, ꓸꓸꓸ), ᐸꟷ(selᴛ7, ꓸꓸꓸ), ᐸꟷ(selᴛ8, ꓸꓸꓸ), ᐸꟷ(selᴛ9, ꓸꓸꓸ))) {
    case 0: {
        fmt.Println("unexpected send to nil channel");
        break;
    }
    case 1 when selᴛ6.ꟷᐳ(out var v1): {
        fmt.Println("unexpected received from nil channel: ", v1);
        break;
    }
    case 2 when selᴛ7.ꟷᐳ(out var v1): {
        fmt.Println("closed channel 2 selected immediately: ", v1);
        break;
    }
    case 3 when selᴛ8.ꟷᐳ(out var v1, out var okΔ2): {
        fmt.Println("unexpected: OK: ", okΔ2, " -- got: ", v1);
        break;
    }
    case 4 when selᴛ9.ꟷᐳ(out a[f()]): {
        fmt.Println("unexpected: ", a[f()]);
        break;
    }}
    var s = new nint[]{7, 2, 8, -9, 4, 0}.slice();
    var c = new channel<nint>(0);
    goǃ(sum, s[..(int)(len(s) / 2)], c);
    goǃ(sum, s[(int)(len(s) / 2)..], c);
    goǃ(sum, s[2..5], c);
    nint x = ᐸꟷ(c);
    nint y = ᐸꟷ(c);
    nint z = ᐸꟷ(c);
    fmt.Println(x, y, x + y, z);
    var fΔ1 = new channel<nint>(0);
    var quit = new channel<nint>(0);
    var fʗ1 = fΔ1;
    var quitʗ1 = quit;
    goǃ(() => {
        for (nint i = 0; i < 10; i++) {
            fmt.Println(ᐸꟷ(fʗ1));
        }
        quitʗ1.ᐸꟷ(0);
    });
    fibonacci(fΔ1, quit);
    var mychanl = new channel<@string>(0);
    goǃ(sendOnly, mychanl);
    var (result, ok) = ᐸꟷ(mychanl, ꟷ);
    fmt.Println(result, ok);
    foreach (var v in range(((IntSlice)s).All())) {
        fmt.Println(v);
    }
    sieve();
    var ca = new channel<@string>(1);
    var cb = new channel<@string>(1);
    ca.ᐸꟷ("hello"u8);
    fmt.Println(firstMsg(ca, cb));
    var done = new channel<EmptyStruct>(0);
    fmt.Println(poll(done));
    close(done);
    fmt.Println(poll(done));
    var (r, outerPrimary) = raceSend();
    fmt.Println("raced:", r.value, r.primary, outerPrimary);
}

internal static @string firstMsg(channel<@string> a, channel<@string> b) {
    var selᴛ10 = a;
    var selᴛ11 = b;
    switch (select(ᐸꟷ(selᴛ10, ꓸꓸꓸ), ᐸꟷ(selᴛ11, ꓸꓸꓸ))) {
    case 0 when selᴛ10.ꟷᐳ(out var m): {
        return "a:"u8 + m;
    }
    case 1 when selᴛ11.ꟷᐳ(out var m): {
        return "b:"u8 + m;
    }}
    return default!;
}

internal static @string poll(channel<EmptyStruct> done) {
    var selᴛ12 = done;
    switch (trySelect(ᐸꟷ(selᴛ12, ꓸꓸꓸ))) {
    case 0 when selᴛ12.ꟷᐳ(out _): {
        return "done"u8;
    }
    default: {
        break;
    }}
    return "pending"u8;
}

[GoType] partial struct raceResult {
    internal nint value;
    internal bool primary;
}

internal static (raceResult, bool) raceSend() {
    var results = new channel<raceResult>(1);
    var done = new channel<EmptyStruct>(0);
    var primary = true;
    var doneʗ1 = done;
    var resultsʗ1 = results;
    var racer = (bool primaryΔ1) => {
        var selᴛ13 = doneʗ1;
        switch (select(resultsʗ1.ᐸꟷ(new raceResult(value: 7, primary: primaryΔ1), ꓸꓸꓸ), ᐸꟷ(selᴛ13, ꓸꓸꓸ))) {
        case 0: {
            break;
        }
        case 1 when selᴛ13.ꟷᐳ(out _): {
            break;
        }}
    };
    racer(false);
    return (ᐸꟷ(results), primary);
}

} // end main_package

File: src/Tests/Behavioral/SelectStatement/SelectStatement.go
package main

import (
	"fmt"
)

type IntSlice []int

func g1(ch chan int) {
	ch <- 12
}

func g2(ch chan int) {
	ch <- 32
}

func sum(s []int, c chan int) {
	sum := 0
	for _, v := range s {
		sum += v
	}
	c <- sum // send sum to c
}

func fibonacci(f, quit chan int) {
	x, y := 0, 1
	for {
		select {
		case f <- x:
			x, y = y, x+y
		case <-quit:
			fmt.Println("quit")
			return
		}
	}
}

func sendOnly(s chan<- string) {
	s <- "output"
}

// All is an iterator over the elements of s.
func (s IntSlice) All() func(yield func(int) bool) {
	return func(yield func(int) bool) {
		for _, v := range s {
			if !yield(v) {
				return
			}
		}
	}
}

// Send the sequence 2, 3, 4, … to channel 'ch'.
func generate(ch chan<- int) {
	for i := 2; ; i++ {
		ch <- i // Send 'i' to channel 'ch'.
	}
}

// Copy the values from channel 'src' to channel 'dst',
// removing those divisible by 'prime'.
func filter(src <-chan int, dst chan<- int, prime int) {
	for i := range src { // Loop over values received from 'src'.
		if i%prime != 0 {
			dst <- i // Send 'i' to channel 'dst'.
		}
	}
}

// The prime sieve: Daisy-chain filter processes together.
func sieve() {
	ch := make(chan int) // Create a new channel.
	go generate(ch)      // Start generate() as a subprocess.
	for {
		prime := <-ch
		fmt.Print(prime, "\n")
		ch1 := make(chan int)
		go filter(ch, ch1, prime)
		ch = ch1

		if prime > 40 {
			break
		}
	}
}

func f() int {
	return 0
}

func main() {
	ch := make(chan int, 2)

	ch <- 1
	ch <- 2

	fmt.Println(<-ch)
	fmt.Println(<-ch)

	a := make([]int, 2)
	ch1 := make(chan int)
	ch2 := make(chan int)
	ch3 := make(chan int)
	ch4 := make(chan int)

	go g1(ch1)
	go g2(ch2)
	go g1(ch3)
	go g2(ch4)

	for i := 0; i < 4; i++ {
		select {
		case v1 := <-ch1:
			fmt.Println("Got: ", v1)
		case v1 := <-ch2:
			fmt.Println("Got: ", v1)
		case v1, ok := <-ch3:
			fmt.Println("OK: ", ok, " -- got: ", v1)
		case a[f()] = <-ch4:
			fmt.Println("Got: ", a[f()])
			//default:
			//    fmt.Println("Default")
		}
	}

	s := []int{7, 2, 8, -9, 4, 0}

	c := make(chan int)
	go sum(s[:len(s)/2], c)
	go sum(s[len(s)/2:], c)
	go sum(s[2:5], c)
	x, y := <-c, <-c // receive from c
	z := <-c
	fmt.Println(x, y, x+y, z)
	f := make(chan int)
	quit := make(chan int)
	go func() {
		for i := 0; i < 10; i++ {
			fmt.Println(<-f)
		}
		quit <- 0
	}()
	fibonacci(f, quit)

	mychanl := make(chan string)

	// function converts bidirectional channel to send only channel
	go sendOnly(mychanl)

	result, ok := <-mychanl
	fmt.Println(result, ok)
	for v := range IntSlice(s).All() {
		fmt.Println(v)
	}
	sieve()
}
